lldb/source/Plugins/Trace/intel-pt/DecodedThread.cpp

```diff
@@ -85,6 +85,11 @@ double DecodedThread::NanosecondsRange::GetInterpolatedTime(
   return interpolate(next_range->nanos);
 }
 
+DecodedThread::TraceItemStorage::TraceItemStorage(
+    const TraceItemStorage &other) {
+  std::memcpy(this, &other, sizeof *this);
```

Comment:
Does the `DecodedThread` know which member of the `TraceItemStorage` is being used? If so, you can pass an additional argument and copy the right member rather than just copying the bytes.

Reply:
`DecodedThread` does know which member is being used (it uses `enum TraceItemKind` to tag the union); however, it stores the tag separately in `DecodedThread::m_item_kinds` for space/alignment reasons.
I added this copy constructor for use by `std::vector<TraceItemStorage> DecodedThread::m_trace_data`; without it, it raises a compile error because it doesn't know how to copy/move `TraceItemStorage` elements when resizing the vector. This means that we can't add a second TraceItemKind argument to the copy constructor, because std::vector<TraceItemStorage> won't know how to use that.

Reply:
I think that the copies might cause unnecessary slowdowns because traces can contain tens of millions of items. What about using a deque instead of a vector? Then you don't need resizes anymore.

Reply:
Thanks for the  suggestion, I changed `DecodedThread::m_item_{kinds,data}` to `std::deque`s and removed the `TraceItemStorage`'s copy constructor.